DDEyC-Assistant/Services/ChatService.cs

```diff
@@ -111,128 +111,200 @@ public async Task<ChatStartResultDto> StartChatAsync(int userId)
     }
 
     public async Task<ChatResponseDto> ProcessChatAsync(ChatRequestDto chatRequest, int userId)
+    {
+        var activeThread = await ValidateAndGetActiveThread(userId, chatRequest.ThreadId);
+        Message userMessage = null;
+        var lockAcquired = false;
+
+        try
+        {
+            lockAcquired = await AcquireConversationLock(chatRequest.ThreadId);
+            await ValidateAndUpdateConversationState(chatRequest.ThreadId);
+
+            var (response, newUserMessage) = await ProcessChatMessageAsync(chatRequest, activeThread, userMessage);
+            userMessage = newUserMessage;
+            return response;
+        }
+        catch (Exception ex)
+        {
+            await HandleChatProcessingException(ex, userMessage?.Id);
+            return null; // This return is unreachable as HandleChatProcessingException always throws
+        }
+        finally
+        {
+            await CleanupChatProcessing(chatRequest.ThreadId, lockAcquired);
+        }
+    }
+
+    private async Task<UserThread> ValidateAndGetActiveThread(int userId, string threadId)
     {
         var activeThread = await _chatRepository.GetActiveThreadForUser(userId);
-        if (activeThread == null || activeThread.ThreadId != chatRequest.ThreadId)
+        if (activeThread == null || activeThread.ThreadId != threadId)
         {
             throw new ChatServiceException("Invalid thread for this user.", "INVALID_THREAD");
         }
+        return activeThread;
+    }
 
-        // Try to acquire lock for this specific conversation
-        if (!await _lockManager.AcquireLock(chatRequest.ThreadId, TimeSpan.FromSeconds(2)))
+    private async Task<bool> AcquireConversationLock(string threadId)
+    {
+        var lockAcquired = await _lockManager.AcquireLock(threadId, TimeSpan.FromSeconds(2));
+        if (!lockAcquired)
         {
             throw new ChatServiceException(
-                "Esta conversación está procesando otro mensaje. Por favor, espere.",
+                "Invalid thread or thread is processing another message.",
                 "CONVERSATION_BUSY");
         }
+        return lockAcquired;
+    }
 
-        Message userMessage = null;
-        try
+    private async Task ValidateAndUpdateConversationState(string threadId)
+    {
+        var conversationState = await _chatRepository.GetConversationState(threadId);
+        if (conversationState?.State == ConversationState.Processing)
         {
-            // Check if there's a pending operation
-            var conversationState = await _chatRepository.GetConversationState(chatRequest.ThreadId);
-            if (conversationState?.State == ConversationState.Processing)
+            if (DateTime.UtcNow - conversationState.LastOperation > _conversationTimeout)
```

Comment:
Let's make sure we're using Unix timestamps here to avoid any issues with time zones.

Reply:
https://www.reddit.com/r/programming/comments/ker87v/timestamp_is_the_right_way_to_handle_different/?rdt=58363

Reply:
https://www.akashmittal.com/mysql-timestamp-right-way/

Reply:
si lo quieres hacer muy exacto se puede usar long unixTimestamp = DateTimeOffset.UtcNow.ToUnixTimeSeconds();